src/UIHandler.cpp

```diff
@@ -0,0 +1,345 @@
+            m_offset.x = ( TERMX > m_thisSize.x ) ? ( TERMX - m_thisSize.x ) / 2 : 0;
+            m_offset.y = ( TERMY > m_thisSize.y ) ? ( TERMY - m_thisSize.y ) / 2 : 0;
+            break;
+
+        default:
+            DebugLog( D_ERROR, DC_ALL ) << "Please state the location...";
+            break;
+    }
+
+    // set up window
+    m_wf_win = newwin( m_thisSize.y, m_thisSize.x, m_offset.y, m_offset.x );
+    m_wf_winptr = WINDOW_PTR( m_wf_win );
+
+    m_lastSize = m_thisSize;
+
+    m_lastLocation = m_thisLocation;
+    return 0;
+}
+
+void UIWindow::DrawEverything()
+{
+    werase( m_wf_win );
+    m_panel->DrawEverything( m_wf_win, { 0, 0 } );
+}
+
+std::vector<std::shared_ptr<UIPanel>> UIPaddingPanel::GetChild() const
+{
+    return m_childPanels;
+}
+
+void UIPaddingPanel::AddChild( std::shared_ptr<UIPanel> panel )
+{
+    if( !m_childPanels.empty() ) {
+        DebugLog( D_ERROR, DC_ALL ) << "Only supports one panel";
+        return;
+    }
+    m_childPanels.push_back( panel );
+}
+
+void UIPaddingPanel::RemoveChild( size_t index )
+{
+    if( index != m_childPanels.size() - 1 ) {
+        m_childPanels[index] = std::move( m_childPanels.back() );
+    }
+    m_childPanels.pop_back();
+}
+
+point UIPaddingPanel::RequestedSize( Sizes sizes )
+{
+    point size;
+
+    if( m_drawBorder )
+        size += { 2, 2 };
+
+    if( m_childPanels.empty() ) {
+        return size;
+    }
+
+    auto paneSize = m_childPanels[0]->RequestedSize( sizes );
+    size += paneSize;
+
+    return size;
+}
+
+// We are a simple border!
+int UIPaddingPanel::SetSize( point size )
+{
+    m_thisSize = size;
+
+    if( m_childPanels.empty() ) {
+        return 0;
+    }
+    auto newSize = size;
+
+    if( m_drawBorder )
+        newSize -= { 2, 2 };
+
+    m_childPanels[0]->SetSize( newSize );
+    return 0;
+}
+
+void UIPaddingPanel::DrawEverything( WINDOW *wf_win, point offset )
+{
+
+    if( m_drawBorder ) {
+        UIUtils::DrawBorder( wf_win, offset, m_thisSize );
+    }
+
+    if( !m_childPanels.empty() ) {
+        point offset;
+
+        if( m_drawBorder )
+            offset += { 1, 1 };
+
+        m_childPanels[0]->DrawEverything( wf_win, offset );
+    }
+
+    m_lastSize = m_thisSize;
+}
+
+
+UIPaddingPanel::UIPaddingPanel( bool drawBorder )
+{
+    m_drawBorder = drawBorder;
+}
+
+std::vector<std::shared_ptr<UIPanel>> UITabPanel::GetChild() const
+{
+    return m_childPanels;
+}
+
+void UITabPanel::AddChild( std::shared_ptr<UIPanel> panel )
+{
+    AddChild( std::to_string( m_childPanels.size() ), panel );
+}
+
+void UITabPanel::AddChild( std::string name, std::shared_ptr<UIPanel> panel )
+{
+    m_childPanels.push_back( panel );
+    m_childPanelNames.push_back( name );
+}
+
+void UITabPanel::RemoveChild( size_t index )
+{
+    if( index != m_childPanels.size() - 1 ) {
+        m_childPanels[index] = std::move( m_childPanels.back() );
+    }
+    m_childPanels.pop_back();
+
+    if( index != m_childPanelNames.size() - 1 ) {
```

Comment:
Ugly code duplication, it's the same removal logic as above for the childPanels vector. Just put the name and the child panel in one vector, maybe as `std::pair<std::string, std::shared_ptr<UIPanel>>`.